lua/entities/gmod_wire_expression2/core/custom/sv_selfaware2.lua

```diff
@@ -1,54 +1,125 @@
-        local fncnameproper = funcname:match("(%w*)%(") or funcname
-		for Name,Data in pairs(funcs) do
-            local proper = Name:match("(%w*)%(")
-            if proper and proper==fncnameproper then return Data[3] end -- Return function runtime
-        end
-    else
-        return e2func[3]
-    end
-end
+-- TODO: Set E2 costs ( __e2setcost(N) ).
 
--- Ex: print(defined("print(...)")) or print(defined("e:health()"))
--- Returns number, 0 being not defined, 1 being defined as an official e2 function, 2 being a user-defined function
+-- Ex: print(defined("print(...)")) or print(defined("health(e:)"))
+-- Returns number, 0 being not defined, 1 being defined as an official e2 function, 2 being a user-defined function.
+-- Note: If you are checking for availability of the builtin function and if you know the signature ahead of time,
+--       then it is preferred to use #ifdef pre-processor statement; this function exists for dynamic kind of checks.
 e2function number defined(string funcname)
-	return getE2Func(self,funcname) and 1 or (getE2UDF(self,funcname) and 2 or 0)
+    -- Check UDF first (see the above note for why).
+    local isUDF, udfDirect = getE2UDF(self, funcname)
+    if udfDirect then return 2 end -- UDF perfect match.
+    local isFunc, funcDirect = getE2Func(self, funcname)
+    if funcDirect then return 1 end -- Builtin perfect match.
+    -- Name only match after this point :(
+    -- Which one to prefer if they are both defined with identical name? It makes sense to prefer UDF?    *smh*
+    -- Should this return a negative number to indicate uncertainity?
+    -- To be honest, this function should be split into 2, respectively. To avoid such dilemma...
+    if isUDF --[[and not isFunc]] then return 2 end -- Found named UDF match.
+    if isFunc --[[and not isUDF]] then return 1 end -- Found named builtin match.
+    return 0
 end
 
--- Ex: print(getFunctionPath("print(...)")) Will return builtins or something, idr where print is actually
+-- Ex: print(getFunctionPath("print(...)")) would print the path to .../core/debug.lua file.
 -- Returns the path where the function was defined, useful for finding whether something was added with an addon.
 e2function string getFunctionPath(string funcname)
-	local func = getE2Func(self,funcname)
-	if not func then return "" end
-	return debug.getinfo(func,"S").short_src
+    local func = getE2Func(self, funcname)
+    -- source is better than short_src, because it can help identify custom addon/core more easily.
+    return isfunction(func) and debug_getinfo(func, "S").source or ""
+end
+
+-- Returns an array containing only names of all User-Defined Functions.
+e2function array udfNames()
+    -- Populate keys on the table (to avoid duplicate entries) and then get the keys on return.
+    local ret = {}
+    for name in pairs(self.funcs_ret) do
+        local idx = string_find(name, "(", 2, true) -- This should never return a nil.
```

Comment:
Not using patterns here because I think this will be faster; the 4th argument `true` disables patterns for performance.